docs/source/cpp/streaming_execution.rst

```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+
+``select_k_sink`` option enables selecting k number of elements. 
+:class:`arrow::compute::SelectKOptions` which is a defined by 
+using :struct:`OrderBySinkNode` definition. This option returns a sink node that receives 
+inputs and then compute top_k/bottom_k.
+Note: This node is a "pipeline breaker" and will fully materialize the input in memory.
+In the future, spillover mechanisms will be added which should alleviate this constraint.
+
+Create SelectK Option::
+
+  arrow::compute::SelectKOptions options = arrow::compute::SelectKOptions::TopKDefault(
+              /*k=*/2, {"i32"});
+
+  ARROW_ASSIGN_OR_RAISE(
+    arrow::compute::ExecNode * k_sink_node,
+    arrow::compute::MakeExecNode("select_k_sink",
+      plan.get(), {source},
+      arrow::compute::SelectKSinkNodeOptions{options, &sink_gen}));
+
+
+SelectK example:
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: KSelect Example)
+  :end-before: (Doc section: KSelect Example)
+  :linenos:
+  :lineno-match:
+
+.. _stream_execution_scan_docs:
+
+``scan``
+---------
+
+`scan` is an operation used to load and process datasets.  It should be preferred over the
+more generic `source` node when your input is a dataset.  The behavior is defined using 
+:class:`arrow::dataset::ScanNodeOptions`.  More information on datasets and the various
+scan options can be found in :ref:`dataset<cpp-dataset-reading>`.
+
+This node is capable of applying pushdown filters to the file readers which reduce
+the amount of data that needs to be read.  This means you may supply the same
+filter expression to the scan node that you also supply to the FilterNode because
+the filtering is done in two different places.
+
+Creating a Scan `ExecNode`::
+
+  auto options = std::make_shared<arrow::dataset::ScanOptions>();
+  options->use_async = true; 
+  options->projection = Materialize({});  // create empty projection
+
+  // construct the scan node
+  cp::ExecNode* scan;
+  auto scan_node_options = arrow::dataset::ScanNodeOptions{dataset, options};
+
+  ARROW_ASSIGN_OR_RAISE(scan,
+                          cp::MakeExecNode("scan", plan.get(), {}, 
+                            scan_node_options));
+
+Scan example:
+
+.. literalinclude:: ../../../cpp/examples/arrow/execution_plan_documentation_examples.cc
+  :language: cpp
+  :start-after: (Doc section: Scan Example)
+  :end-before: (Doc section: Scan Example)
+  :linenos:
+  :lineno-match:
+
+.. _stream_execution_hashjoin_docs:
+
+``hash_join``
+-------------
+
+``hash_join`` operation provides the relational algebra operation, join using hash-based
+algorithm. :class:`arrow::compute::HashJoinNodeOptions` contains the options required in 
+defining a join. JoinType can be one of LEFT_SEMI, RIGHT_SEMI, LEFT_ANTI, RIGHT_ANTI,
+INNER, LEFT_OUTER, RIGHT_OUTER and FULL_OUTER. Also the join-key or by which column/s, 
+and output suffixes can be set via the the join options. 
```

Comment:
What is an output suffix?

Reply:
Perhaps 

"The hash_join supports left/right/full semi/anti/outerjoins. 
Also the join-key (i.e. the column(s) to join on), and suffixes (i.e a suffix term like "_x"
which can be appended as a suffix for column names duplicated in both left and right 
relations.) can be set via the the join options." ?

Reply:
Thanks, I'd say we should explain that in a little more detail (or just omit it) since I don't think this is a standard feature of joins (at least in SQL, I may be wrong)

Reply:
Let me add a little more details.
Is it okay if we include external resources like peer reviewed papers, wiki pages, standard docs as helper materials?

Reply:
Yes, totally fine. (And I might just be ignorant of this if it's a common option.)

Reply:
Great, I will add more details to this.

Reply:
References and links added.